walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,103 @@
+
+        Collection<TimeRange> availableTimes = new ArrayList<>();
+        Collection<TimeRange> optionalTimes = new ArrayList<>();
+        Collection<Event> optionalOnlyEvents = new ArrayList<>();
+
+        if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+            return availableTimes;
+        }
+        if (events.isEmpty()) {
+            availableTimes.add(TimeRange.WHOLE_DAY);
+            return availableTimes;
+        }
+
+        for (Event event : events) {
+            TimeRange eventRange = event.getWhen();
+            Set<String> eventAttendees = event.getAttendees();
+
+            boolean hasNoRequiredAttendees = Collections.disjoint(mandatoryAttendees, eventAttendees);
+            // are any optional attendees attending this event
+            boolean hasOnlyOptionalAttendees = !(Collections.disjoint(optionalAttendees, eventAttendees));
+             
+            // if no mandatory attendees are at the event
+            if (hasNoRequiredAttendees) {
+                // if only optional attendees are at the event
+                if (hasOnlyOptionalAttendees) {
+                    optionalOnlyEvents.add(event);
+                } else if (availableTimes.isEmpty()) {
+                    availableTimes.add(TimeRange.WHOLE_DAY);
+                }
+                continue;
+            }
+
+            // creates TimeRange available before and after event
+            TimeRange availableBefore = 
+                TimeRange.fromStartEnd(TimeRange.START_OF_DAY, eventRange.start(), /* inclusive= */ false);
+            TimeRange availableAfter = 
+                TimeRange.fromStartEnd(eventRange.end(), TimeRange.END_OF_DAY, /* inclusive= */ true);
+            
+            availableTimes = returnOverlaps(availableBefore, availableAfter, availableTimes, meetingDuration);
+        }
+
+        for (Event optionalEvent : optionalOnlyEvents) {
+            TimeRange eventRange = optionalEvent.getWhen();
+
+            TimeRange availableBefore = 
+                TimeRange.fromStartEnd(TimeRange.START_OF_DAY, eventRange.start(), /* inclusive= */ false);
+            TimeRange availableAfter = 
+                TimeRange.fromStartEnd(eventRange.end(), TimeRange.END_OF_DAY, /* inclusive= */ true);
+
+            if (!(optionalTimes.isEmpty()) || mandatoryAttendees.isEmpty()) {
+                optionalTimes = returnOverlaps(availableBefore, availableAfter, optionalTimes, meetingDuration);
+            } else {
+                optionalTimes = returnOverlaps(availableBefore, availableAfter, availableTimes, meetingDuration);
+            }
+        }
+        
+        if (optionalTimes.isEmpty()) {
+            return availableTimes;
+        } else {
+            return optionalTimes;
+        }
+    }
+
+    private Collection<TimeRange> returnOverlaps(TimeRange beforeTimeRange, TimeRange afterTimeRange, Collection<TimeRange> availableTimes, long meetingDuration) {
```

Comment:
Could you format this to make sure your lines are no longer than 80 chars? The editing tool you're using should have some kind of setting/auto-formatter that makes it easy